go/host/l1/publisher.go

```diff
@@ -268,69 +239,67 @@ func (p *Publisher) FetchLatestPeersList() ([]string, error) {
 	return filteredHostAddresses, nil
 }
 
-// `tries` is the number of times to attempt broadcasting the transaction.
-// if awaitReceipt is true then this method will block and synchronously wait to check the receipt, otherwise it is fire
-// and forget and the receipt tracking will happen in a separate go-routine
-func (p *Publisher) signAndBroadcastL1Tx(tx types.TxData, tries uint64, awaitReceipt bool) error {
-	var err error
-	tx, err = p.ethClient.EstimateGasAndGasPrice(tx, p.hostWallet.Address())
-	if err != nil {
-		return errors.Wrap(err, "could not estimate gas/gas price for L1 tx")
-	}
-
-	signedTx, err := p.hostWallet.SignTransaction(tx)
-	if err != nil {
-		return err
-	}
-
-	p.logger.Info("Host issuing l1 tx", log.TxKey, signedTx.Hash(), "size", signedTx.Size()/1024)
-
-	err = retry.Do(func() error {
-		return p.ethClient.SendTransaction(signedTx)
-	}, retry.NewDoublingBackoffStrategy(time.Second, tries)) // doubling retry wait (3 tries = 7sec, 7 tries = 63sec)
-	if err != nil {
-		return fmt.Errorf("could not broadcast L1 tx after %d tries: %w", tries, err)
-	}
-	p.logger.Info("Successfully submitted tx to L1", "txHash", signedTx.Hash())
-
-	if awaitReceipt {
-		// block until receipt is found and then return
-		return p.waitForReceipt(signedTx.Hash())
-	}
+// publishTransaction will keep trying unless the L1 seems to be unavailable or the tx is otherwise rejected
+// It is responsible for keeping the nonce accurate, according to the following rules:
+// - Caller should not increment the wallet nonce before this method is called
+// - This method will increment the wallet nonce only if the transaction is successfully broadcast
+// - This method will continue to resend the tx using latest gas price until it is successfully broadcast or the L1 is unavailable/this service is shutdown
+// - **ONLY** the L1 publisher service is publishing transactions for this wallet (to avoid nonce conflicts)
+func (p *Publisher) publishTransaction(tx types.TxData) error {
```

Comment:
as an improvement for the future, this could receive a cancelable context, so you can cancel the tx from the caller if circumstances change (for ex when it's time to build another rollup).
This could be a todo

Reply:
Yeah good shout, we should have that control, added todo